### PR TITLE
Added business logic for Orders, OrderProducts, Discounts, Cart and o…

### PR DESCRIPTION
**Pull request by IvanVillani**
…thers. Implemented role logic and dependencies, worked on controllers, some other changes and views.

### CODE REVIEW

File: src/main/java/com/ivan/salesapp/services/DiscountService.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        this.discountRepository = discountRepository;
+        this.iProductService = iProductService;
+        this.modelMapper = modelMapper;
+    }
+
+    @Override
+    public List<DiscountServiceModel> findAllDiscounts() {
+        return this.discountRepository.findAll().stream()
+                .map(o -> this.modelMapper.map(o, DiscountServiceModel.class))
+                .collect(toList());
```
Comment (denislav98) [inline src/main/java/com/ivan/salesapp/services/DiscountService.java:43]: format this on new lines

File: src/main/java/com/ivan/salesapp/services/UserService.java [hunk 2]
```diff
@@ -96,20 +123,57 @@ public void setUserRole(String id, String role) {
...
+                    .map(mapToViewModelSetCategories(this.modelMapper))
+                    .collect(toList());
+        }else if ("ROLE_ROOT".equals(authority)){
+            return this.findAllUsers()
+                    .stream()
+                    .map(mapToViewModelSetCategories(this.modelMapper))
+                    .collect(toList());
+        }else{
+            return null;
+        }
```
Comment (denislav98) [inline src/main/java/com/ivan/salesapp/services/UserService.java:160]: 1.Too complex method.
2.It, not a good practice to return null.
3.Think about factory method pattern(In order to deal with different roles)
4.It`s not necessary to use lambda expressions every time. Keep it simple for this one -   
 return this.findAllUsers()
                    .stream()
                    .filter(u -> new ArrayList<>(u.getAuthorities()).get(0).getAuthority().equals("ROLE_MODERATOR"))
                    .map(mapToViewModelSetCategories(this.modelMapper))
                    .collect(toList());
you can implement something like 
Collection<Users> users = this.findAllUsers();

  return getUsersForAuthority(users ,authority);

add an implementation for this method getUsersForAuthority(users ,authority);

File: src/main/java/com/ivan/salesapp/services/UserService.java [hunk 3]
```diff
@@ -96,20 +123,57 @@ public void setUserRole(String id, String role) {
...
+                userServiceModel.getAuthorities().add(this.iRoleService.findByAuthority("ROLE_ADMIN"));
                 break;
         }
 
         this.userRepository.saveAndFlush(this.modelMapper.map(userServiceModel, User.class));
     }
 
+    @Override
+    public List<UserAllViewModel> getUsersBasedOnAuthority(String authority) {
+        if ("ROLE_MODERATOR".equals(authority)){
```
Comment (denislav98) [inline src/main/java/com/ivan/salesapp/services/UserService.java:141]: create Enum class for this roles

File: src/main/java/com/ivan/salesapp/services/UserService.java [hunk 4]
```diff
@@ -96,20 +123,57 @@ public void setUserRole(String id, String role) {
...
+            } else {
+                user.setAuthorities(u.getAuthorities()
+                        .stream()
+                        .map(RoleServiceModel::getAuthority)
+                        .map(s -> s.substring(5))
+                        .collect(toSet()));
+            }
+            return user;
+        };
+    }
```
Comment (denislav98) [inline src/main/java/com/ivan/salesapp/services/UserService.java:178]: if(userHasAuthority){
 doSmth or call method that returns User
}else{
   doSmth or call method that returns User

}

File: src/main/java/com/ivan/salesapp/web/controllers/CartController.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        return (List<ShoppingCartItem>) session.getAttribute("shopping-cart");
+    }
+
+    private void initCart(HttpSession session) {
+        if (session.getAttribute("shopping-cart") == null) {
+            session.setAttribute("shopping-cart", new LinkedList<>());
+        }
+    }
+
+    private void addItemToCart(ShoppingCartItem item, List<ShoppingCartItem> cart) {
```
Comment (denislav98) [inline src/main/java/com/ivan/salesapp/web/controllers/CartController.java:104]: rename cart to cartItems

File: src/main/java/com/ivan/salesapp/web/controllers/CartController.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        for (ShoppingCartItem shoppingCartItem : cart) {
+            if (shoppingCartItem.getProduct().getProduct().getId().equals(item.getProduct().getProduct().getId())) {
+                shoppingCartItem.setQuantity(shoppingCartItem.getQuantity() + item.getQuantity());
+                return;
+            }
```
Comment (denislav98) [inline src/main/java/com/ivan/salesapp/web/controllers/CartController.java:109]: remove return and just add else statement 
if(..){
                shoppingCartItem.setQuantity(shoppingCartItem.getQuantity() + item.getQuantity());
}else{
        cart.add(item);

}

File: src/main/java/com/ivan/salesapp/web/controllers/CartController.java [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        }
+
+        cart.add(item);
+    }
+
+    private void removeItemFromCart(String id, List<ShoppingCartItem> cart) {
+        cart.removeIf(ci -> ci.getProduct().getProduct().getId().equals(id));
```
Comment (denislav98) [inline src/main/java/com/ivan/salesapp/web/controllers/CartController.java:116]: cartItem.removeIf(item -> item.geProduct()...

File: src/main/java/com/ivan/salesapp/services/UserService.java (hunk 2 above)
Comment (IvanVillani) [reply src/main/java/com/ivan/salesapp/services/UserService.java:160]: Actually this method wasn't supposed to be like this, I forgot to fix it and made the pull request.
On the next day I saw it and fixed it. Sorry for this dumb mistake :)

**Pull request closed without merging**